File: project/ProjectFinal/backend/MedicalAppointments.API/Services/NotificationService.cs
using MailKit.Net.Smtp;
using MedicalAppointments.API.Infrastructure;
using MedicalAppointments.API.Models;
using MimeKit;

namespace MedicalAppointments.API.Services;

public interface INotificationService
{
    Task SendWaitingListNotificationAsync(string email, string name, TimeSlot slot);
    Task SendPenaltyNotificationAsync(User patient);
    Task SendBookingConfirmationAsync(User patient, Appointment appt, TimeSlot slot);
    Task SendCancellationConfirmationAsync(User patient, TimeSlot slot);
}

public class NotificationService : INotificationService
{
    private readonly IConfiguration _config;
    private readonly AppDbContext   _db;
    private readonly ILogger<NotificationService> _logger;

    public NotificationService(
        IConfiguration config,
        AppDbContext db,
        ILogger<NotificationService> logger)
    {
        _config = config;
        _db     = db;
        _logger = logger;
    }

    public async Task SendWaitingListNotificationAsync(string email, string name, TimeSlot slot)
    {
        var subject = "🗓️ Un créneau vient de se libérer !";
        var body = $@"
            <div style='font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;'>
                <h2 style='color: #2563eb;'>Bonne nouvelle, {name} !</h2>
                <p>Un créneau vient de se libérer :</p>
                <div style='background: #f0f9ff; border-left: 4px solid #2563eb; padding: 16px; margin: 16px 0;'>
                    <strong>📅 Date :</strong> {slot.StartTime:dddd dd MMMM yyyy}<br/>
                    <strong>🕐 Heure :</strong> {slot.StartTime:HH:mm} – {slot.EndTime:HH:mm}
                </div>
                <p>Connectez-vous rapidement pour le réserver avant qu'il ne soit pris !</p>
                <p style='color: #6b7280; font-size: 12px;'>Ce message a été envoyé car vous êtes inscrit(e) sur la liste d'attente.</p>
            </div>";

        await SendEmailAsync(email, subject, body);
    }

    public async Task SendPenaltyNotificationAsync(User patient)
    {
        var subject = "⚠️ Pénalité appliquée sur votre compte";
        var body = $@"
            <div style='font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;'>
                <h2 style='color: #dc2626;'>Pénalité activée</h2>
                <p>Bonjour {patient.FullName},</p>
                <p>Suite à l'annulation de <strong>{patient.CancelCount} rendez-vous</strong>, 
                   votre compte a été temporairement restreint.</p>
                <div style='background: #fef2f2; border-left: 4px solid #dc2626; padding: 16px; margin: 16px 0;'>
                    <strong>Blocage jusqu'au :</strong> {patient.PenaltyUntil:dd/MM/yyyy à HH:mm}
                </div>
                <p>Pendant cette période, vous ne pourrez pas prendre de nouveaux rendez-vous.</p>
                <p>Pour toute question, contactez notre secrétariat.</p>
            </div>";

        await SendEmailAsync(patient.Email, subject, body);
    }

    public async Task SendBookingConfirmationAsync(User patient, Appointment appt, TimeSlot slot)
    {
        var subject = "✅ Confirmation de votre rendez-vous";
        var cancelInfo = appt.AnonymousToken != null
            ? $@"<div style='background: #fffbeb; border-left: 4px solid #f59e0b; padding: 16px; margin: 16px 0;'>
                    <strong>⚠️ Code d'annulation :</strong> <code style='font-size: 16px; font-weight: bold;'>{appt.AnonymousToken}</code><br/>
                    <small>Conservez ce code précieusement pour pouvoir annuler votre rendez-vous.</small>
                </div>"
            : "<p>Pour annuler, connectez-vous à votre espace patient.</p>";

        var body = $@"
            <div style='font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;'>
                <h2 style='color: #16a34a;'>Rendez-vous confirmé ✓</h2>
                <p>Bonjour {patient.FullName},</p>
                <p>Votre rendez-vous a bien été enregistré :</p>
                <div style='background: #f0fdf4; border-left: 4px solid #16a34a; padding: 16px; margin: 16px 0;'>
                    <strong>📅 Date :</strong> {slot.StartTime:dddd dd MMMM yyyy}<br/>
                    <strong>🕐 Heure :</strong> {slot.StartTime:HH:mm} – {slot.EndTime:HH:mm}<br/>
                    {(appt.Reason != null ? $"<strong>📝 Motif :</strong> {appt.Reason}" : "")}
                </div>
                {cancelInfo}
            </div>";

        await SendEmailAsync(patient.Email, subject, body);
    }

    public async Task SendCancellationConfirmationAsync(User patient, TimeSlot slot)
    {
        var subject = "❌ Annulation de votre rendez-vous";
        var body = $@"
            <div style='font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;'>
                <h2 style='color: #dc2626;'>Rendez-vous annulé</h2>
                <p>Bonjour {patient.FullName},</p>
                <p>Votre rendez-vous du <strong>{slot.StartTime:dddd dd MMMM yyyy à HH:mm}</strong> 
                   a bien été annulé.</p>
                <p>Vous pouvez prendre un nouveau rendez-vous à tout moment.</p>
            </div>";

        await SendEmailAsync(patient.Email, subject, body);
    }

    // ─── Envoi SMTP via MailKit ──────────────────────────────────────────
    private async Task SendEmailAsync(string to, string subject, string htmlBody)
    {
        // Persister la notification dans la DB avant l'envoi
        var notification = new Notification
        {
            RecipientEmail = to,
            Subject        = subject,
            Body           = htmlBody,
            Status         = NotificationStatus.Pending
        };
        _db.Notifications.Add(notification);
        await _db.SaveChangesAsync();

        try
        {
            var message = new MimeMessage();
            message.From.Add(MailboxAddress.Parse(_config["Email:From"]));
            message.To  .Add(MailboxAddress.Parse(to));
            message.Subject = subject;
            message.Body    = new TextPart("html") { Text = htmlBody };

            using var smtp = new SmtpClient();
            await smtp.ConnectAsync(
                _config["Email:Host"],
                int.Parse(_config["Email:Port"]!),
                MailKit.Security.SecureSocketOptions.StartTls);
            await smtp.AuthenticateAsync(_config["Email:User"], _config["Email:Pass"]);
            await smtp.SendAsync(message);
            await smtp.DisconnectAsync(true);

            notification.Status = NotificationStatus.Sent;
            notification.SentAt = DateTime.UtcNow;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Échec envoi email vers {Email}", to);
            notification.Status = NotificationStatus.Failed;
        }
        finally
        {
            await _db.SaveChangesAsync();
        }
    }
}

File: appointment-service/src/main/java/com/hospital/appointment/service/NotificationService.java
package com.hospital.appointment.service;

import com.hospital.appointment.entity.*;
import com.hospital.appointment.repository.NotificationRepository;
import jakarta.mail.internet.MimeMessage;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

/**
 * Service d'envoi de notifications email + persistance en BDD.
 * Migré depuis Services/NotificationService.cs (.NET).
 *
 * Utilise JavaMailSender (équivalent MailKit côté .NET).
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class NotificationService {

    private final JavaMailSender mailSender;
    private final NotificationRepository notificationRepository;

    @Value("${app.email.from}")
    private String fromAddress;

    private static final DateTimeFormatter DATE_FR =
            DateTimeFormatter.ofPattern("EEEE dd MMMM yyyy", Locale.FRENCH);
    private static final DateTimeFormatter TIME_FR =
            DateTimeFormatter.ofPattern("HH:mm");
    private static final DateTimeFormatter DATETIME_FR =
            DateTimeFormatter.ofPattern("dd/MM/yyyy 'à' HH:mm", Locale.FRENCH);

    public void sendWaitingListNotification(String email, String name, TimeSlot slot) {
        String subject = "🗓️ Un créneau vient de se libérer !";
        String body = """
                <div style='font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;'>
                    <h2 style='color: #2563eb;'>Bonne nouvelle, %s !</h2>
                    <p>Un créneau vient de se libérer :</p>
                    <div style='background: #f0f9ff; border-left: 4px solid #2563eb; padding: 16px; margin: 16px 0;'>
                        <strong>📅 Date :</strong> %s<br/>
                        <strong>🕐 Heure :</strong> %s – %s
                    </div>
                    <p>Connectez-vous rapidement pour le réserver avant qu'il ne soit pris !</p>
                    <p style='color: #6b7280; font-size: 12px;'>Ce message a été envoyé car vous êtes inscrit(e) sur la liste d'attente.</p>
                </div>
                """.formatted(name,
                slot.getStartTime().format(DATE_FR),
                slot.getStartTime().format(TIME_FR),
                slot.getEndTime().format(TIME_FR));

        sendEmail(email, subject, body);
    }

    public void sendPenaltyNotification(User patient) {
        String subject = "⚠️ Pénalité appliquée sur votre compte";
        String body = """
                <div style='font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;'>
                    <h2 style='color: #dc2626;'>Pénalité activée</h2>
                    <p>Bonjour %s,</p>
                    <p>Suite à l'annulation de <strong>%d rendez-vous</strong>,
                       votre compte a été temporairement restreint.</p>
                    <div style='background: #fef2f2; border-left: 4px solid #dc2626; padding: 16px; margin: 16px 0;'>
                        <strong>Blocage jusqu'au :</strong> %s
                    </div>
                    <p>Pendant cette période, vous ne pourrez pas prendre de nouveaux rendez-vous.</p>
                    <p>Pour toute question, contactez notre secrétariat.</p>
                </div>
                """.formatted(patient.getFullName(),
                patient.getCancelCount(),
                patient.getPenaltyUntil() != null ? patient.getPenaltyUntil().format(DATETIME_FR) : "—");

        sendEmail(patient.getEmail(), subject, body);
    }

    public void sendBookingConfirmation(User patient, Appointment appt, TimeSlot slot) {
        String subject = "✅ Confirmation de votre rendez-vous";

        String cancelInfo;
        if (appt.getAnonymousToken() != null) {
            cancelInfo = """
                    <div style='background: #fffbeb; border-left: 4px solid #f59e0b; padding: 16px; margin: 16px 0;'>
                        <strong>⚠️ Code d'annulation :</strong> <code style='font-size: 16px; font-weight: bold;'>%s</code><br/>
                        <small>Conservez ce code précieusement pour pouvoir annuler votre rendez-vous.</small>
                    </div>
                    """.formatted(appt.getAnonymousToken());
        } else {
            cancelInfo = "<p>Pour annuler, connectez-vous à votre espace patient.</p>";
        }

        String reasonLine = appt.getReason() != null
                ? "<strong>📝 Motif :</strong> " + appt.getReason()
                : "";

        String body = """
                <div style='font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;'>
                    <h2 style='color: #16a34a;'>Rendez-vous confirmé ✓</h2>
                    <p>Bonjour %s,</p>
                    <p>Votre rendez-vous a bien été enregistré :</p>
                    <div style='background: #f0fdf4; border-left: 4px solid #16a34a; padding: 16px; margin: 16px 0;'>
                        <strong>📅 Date :</strong> %s<br/>
                        <strong>🕐 Heure :</strong> %s – %s<br/>
                        %s
                    </div>
                    %s
                </div>
                """.formatted(patient.getFullName(),
                slot.getStartTime().format(DATE_FR),
                slot.getStartTime().format(TIME_FR),
                slot.getEndTime().format(TIME_FR),
                reasonLine,
                cancelInfo);

        sendEmail(patient.getEmail(), subject, body);
    }

    public void sendCancellationConfirmation(User patient, TimeSlot slot) {
        String subject = "❌ Annulation de votre rendez-vous";
        String body = """
                <div style='font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;'>
                    <h2 style='color: #dc2626;'>Rendez-vous annulé</h2>
                    <p>Bonjour %s,</p>
                    <p>Votre rendez-vous du <strong>%s à %s</strong>
                       a bien été annulé.</p>
                    <p>Vous pouvez prendre un nouveau rendez-vous à tout moment.</p>
                </div>
                """.formatted(patient.getFullName(),
                slot.getStartTime().format(DATE_FR),
                slot.getStartTime().format(TIME_FR));

        sendEmail(patient.getEmail(), subject, body);
    }

    /**
     * Persiste la notification, puis tente l'envoi SMTP.
     * Migré depuis SendEmailAsync (.NET) : même logique de tracking en DB.
     */
    @Transactional
    public void sendEmail(String to, String subject, String htmlBody) {
        Notification notification = Notification.builder()
                .recipientEmail(to)
                .subject(subject)
                .body(htmlBody)
                .status(NotificationStatus.Pending)
                .build();
        notificationRepository.save(notification);

        try {
            MimeMessage mimeMessage = mailSender.createMimeMessage();
            MimeMessageHelper helper = new MimeMessageHelper(mimeMessage, "UTF-8");
            helper.setFrom(fromAddress);
            helper.setTo(to);
            helper.setSubject(subject);
            helper.setText(htmlBody, true);   // true = HTML

            mailSender.send(mimeMessage);

            notification.setStatus(NotificationStatus.Sent);
            notification.setSentAt(LocalDateTime.now());
        } catch (Exception ex) {
            log.error("Échec envoi email vers {} : {}", to, ex.getMessage());
            notification.setStatus(NotificationStatus.Failed);
        } finally {
            notificationRepository.save(notification);
        }
    }
}
